satellite/audit/service.go

```diff
@@ -58,18 +62,74 @@ func NewService(log *zap.Logger, config Config, metainfo *metainfo.Service,
 	}, nil
 }
 
+// ReservoirService is a temp name for the service struct during the audit 2.0 refactor.
+// Once V3-2363 and V3-2364 are implemented, Service2 will replace the existing Service struct.
+type ReservoirService struct {
+	log *zap.Logger
+
+	reservoirSlots int
+	Reservoirs     map[storj.NodeID]*Reservoir
+	rand           *rand.Rand
+
+	MetainfoLoop  *metainfo.Loop
+	ReservoirLoop sync2.Cycle
+}
+
+// NewReservoirService instantiates Service2
+func NewReservoirService(log *zap.Logger, metaLoop *metainfo.Loop, r *rand.Rand, config Config) (*ReservoirService, error) {
+	return &ReservoirService{
+		log: log,
+
+		reservoirSlots: config.Slots,
+		Reservoirs:     make(map[storj.NodeID]*Reservoir),
+		rand:           r,
+
+		MetainfoLoop:  metaLoop,
+		ReservoirLoop: *sync2.NewCycle(config.Interval),
+	}, nil
+}
+
+// Run runs auditing service 2.0
+func (service *ReservoirService) Run(ctx context.Context) (err error) {
+	defer mon.Task()(&ctx)(&err)
+	service.log.Info("audit 2.0 is starting up")
+
+	group, ctx := errgroup.WithContext(ctx)
+
+	service.ReservoirLoop.Start(ctx, group, func(ctx context.Context) (err error) {
+		defer mon.Task()(&ctx)(&err)
+		pathCollector := NewPathCollector(service.reservoirSlots, service.rand)
+		err = service.MetainfoLoop.Join(ctx, pathCollector)
+		if err != nil {
+			service.log.Error("error joining metainfoloop", zap.Error(err))
+			return nil
+		}
+		for nodeID, res := range pathCollector.Reservoirs {
+			service.Reservoirs[nodeID] = res
+		}
```

Comment:
What happens if `pathCollector.Reservoirs` does not contain node id that is already present `service.Reservoir`? Will this node id leak in `service.Reservoir`?

Why don't we just make `service.Reservoirs = pathCollector.Reservoirs`?

Reply:
I thought it wasn't good to do `service.Reservoirs = pathCollector.Reservoirs` because the map is a reference type, so service.Reservoirs would just be pointing to the map at pathCollector.Reservoirs. But if you think that's not a big deal then I'll change it.

Reply:
This makes sense. In such case, start with empty `service.Reservoirs` before the for-loop, i.e. `service.Reservoirs = make(map[storj.NodeID]*Reservoir, len(pathCollector.Reservoirs))`. This way we'll avoid leaking reservoirs for nodes that are not present in the latest path collector.

Reply:
Since `pathCollector` is ephemeral, there's no problem in moving the map from one type to another. If the collector was reused then it could be a problem.